src/prefect/utilities/asyncutils.py

```diff
@@ -217,7 +217,24 @@ def sync_compatible(async_fn: T) -> T:
 
     @wraps(async_fn)
     def coroutine_wrapper(*args, **kwargs):
-        if in_async_main_thread():
+        import threading
+
+        from prefect._internal.concurrency.api import create_call, from_sync
+        from prefect._internal.concurrency.runtime import get_runtime_thread
+        from prefect._internal.concurrency.supervisors import get_supervisor
+
+        supervisor = get_supervisor()
+        runtime_thread = get_runtime_thread()
+        current_thread = threading.current_thread()
+        if current_thread.ident == runtime_thread.ident:
+            # In the prefect async context; return the coro for us to await
+            return async_fn(*args, **kwargs)
+        elif supervisor and not is_async_fn(supervisor._call.fn):
+            # In a supervised call that is not async; send the async call to the parent
+            return from_sync.send_call_to_supervising_thread(
+                create_call(async_fn, *args, **kwargs)
+            ).result()
```

Comment:
This is kind of shimmed in here for now since we use `sync_compatible` all over the place. The plan is to rewrite all this once there's a clearer story.